docs/skopeo-inspect.1.md

```diff
@@ -53,9 +61,9 @@ Registry token for accessing the registry.
 
 The number of times to retry; retry wait time will be exponentially increased based on the number of failed attempts.
 
-**--tls-verify**
```

Comment:
Hum, so this exposes a mess (in several subcommands).

The previous CLI framework had a strong distinction between global options and local options, and required the order `skopeo $global-options $command $local-options`. In that context, `--tls-verify` originally was a global option, but became a local per-command option (for _some_ commands, it’s local precisely because `skopeo copy` etc. have separate `--src-tls-verify` and `--dest-tls-verify`), and we have left around global `--tls-verify` for compatibility, but hidden and with a warning.

Cobra has basically no distinction between global/local options; the `tis-verify` option name ends up sharing (some) state. So now we have the global option (in `createApp`) still marked as hidden, so that it doesn’t show up in help in `skopeo copy`, but that hidden bit is shared, and we don’t list it in help of `skopeo inspect` and others.

IIRC I have looked into this some time ago and there wasn’t a clean way to fix this, the `tis-verify` option is either hidden everywhere, or not. It might be worth another look… or, sadly, we might need an exception in the checker.

Reply:
>--tls-verify originally was a global option, but became a local per-command option

Ahhh, thank you, that is important context that helps me understand. I think the important factor here is that `--help` and man pages should match. Right now `--tls-verify` does not appear in any `skopeo help` output, neither globally nor locally, even in `login` which appears to override with a local `tls-verify` option. If we add an exception in the code, that would mean we allow `--tls-verify` to appear in some man pages without appearing in `--help`. That makes me really uncomfortable. I think this needs deeper thought...

Reply:
A possible path would be to not mark `--tls-verify` as hidden in Cobra, but “just” filter it out put the main `skopeo --help` output. I don’t know exactly what this would take but from a quick look it seems possible in principle (`SetHelpFunc`).

That’s extremely ugly but it is _an_ option, if we can’t find anything better.

Reply:
Can't we just mark it hidden for individual commands instead of globally?

Reply:
AFAIK no, because
- The “hidden” bit is shared; so marking the “local” variant of `inspect --tls-verify` hidden would still mark the global one as hidden
- The flag data is basically data-only; we build all subcommand structures in advance, and AFAIK don’t get a change to _change_ what the flags are after Cobra recognizes a subcommand (and for maintainability I would prefer to steer away from doing so anyway)

Reply:
> A possible path would be to not mark `--tls-verify` as hidden in Cobra, but “just” filter it out of the main `skopeo --help` output. I don’t know exactly what this would take but from a quick look it seems possible in principle (`SetHelpFunc`).
> 
> That’s extremely ugly but it is _an_ option, if we can’t find anything better.

That might not actually work either; I can’t get `--tls-verify` to show up in `skopeo inspect --help`.

https://github.com/containers/skopeo/pull/1358 to show what it takes, and maybe someone can figure out another way to get this done. (That PR is already pretty close to unacceptably ugly, even with tests to verify this keeps working.)

Reply:
Alternatively, _if_ Cobra really doesn’t care about option positioning, we could just drop the global option and have existing commands add a local `--tls-verify`, hidden or not and warning or not, depending on circumstances (and we wouldn’t add it to future commands that shouldn’t have it.) That might be quite clean, after all.

Reply:
It's more subtle than that: Cobra does care, but only in one direction: it can accept a global option on the right-hand side, but cannot accept a local (right-hand) option on the left (before the subcommand). Moving `--tls-verify` to a local per-subcommand option would (hypothetically) break legacy tools that run `skopeo --tls-verify=...`. I assume that's the reason for keeping it global and deprecated.

Reply:
Yes, that’s the reason. So much for that idea.

Reply:
Here's a wild idea:
```diff
diff --git a/cmd/skopeo/main.go b/cmd/skopeo/main.go
index 670260e0..4f959932 100644
--- a/cmd/skopeo/main.go
+++ b/cmd/skopeo/main.go
@@ -55,7 +55,7 @@ func createApp() (*cobra.Command, *globalOptions) {
        var dummyVersion bool
        rootCommand.Flags().BoolVarP(&dummyVersion, "version", "v", false, "Version for Skopeo")
        rootCommand.PersistentFlags().BoolVar(&opts.debug, "debug", false, "enable debug output")
-       flag := optionalBoolFlag(rootCommand.PersistentFlags(), &opts.tlsVerify, "tls-verify", "Require HTTPS and verify certificates when accessing the registry")
+       flag := optionalBoolFlag(rootCommand.PersistentFlags(), &opts.tlsVerify, "tls-verifyx", "Require HTTPS and verify certificates when accessing the registry")
        flag.Hidden = true
        rootCommand.PersistentFlags().StringVar(&opts.policyPath, "policy", "", "Path to a trust policy file")
        rootCommand.PersistentFlags().BoolVar(&opts.insecurePolicy, "insecure-policy", false, "run the tool without any policy check")
```
(that is: append 'x' to 'tls-verify'). In my testing, that works to the extent that:
* skopeo --help does not list the option
* skopeo login --help **does** list the option (because it explicitly declares it, I don't know why)
* skopeo --tls-verify=false inspect, delete, list-tags seem to recognize `--tls-verify` as a shortcut for `--tls-verifyx`, and emit the expected warnings (I did not confirm if the option actually works as intended). And
* skopeo --tls-verify=falce copy, sync, standalone*, those don't work at all. I don't know why.

Anhow, seems worth playing around with?

Reply:
I’ll also ask Cobra maintainers, maybe it’s something that could be enabled with a reasonable code change…

Reply:
> (that is: append 'x' to 'tls-verify'). In my testing, that works to the extent that:

…
> * skopeo --tls-verify=false inspect, delete, list-tags seem to recognize `--tls-verify` as a shortcut for `--tls-verifyx`, and emit the expected warnings (I did not confirm if the option actually works as intended). 

I don’t see any warnings in this case (right now, with Cobra 1.2.1 pflag 1.0.5).

Also AFAICS prefixes aren’t matched, as a rule: `skopeo --hel`, `skopeo --override-o inspect` just fail for me.

> * skopeo --tls-verify=falce copy, sync, standalone*, those don't work at all. I don't know why.

That’s consistent with no prefix matching: After the global `--tls-verify` is renamed, there isn’t any `--tls-verify` option for these subcommands, so it seems expected to me that this would fail.

---

In my experiments, the warning actually doesn’t work as expected. It does trigger with `skopeo copy` (when there isn’t a local option, the global option is found and triggers a warning), but both `skopeo inspect --tls=verify=false` and `skopeo --tls-verify=false inspect` edit the local option, and no warning is triggered. (I think this used to work, I’ll look into when it broke.)

So the duplication doesn’t buy us that much; we could just as well have only the global option, and then selectively warn on it based on subcommand choice. (OTOH that doesn’t do what we _truly_ want, which is for the option to only show up in `--help` of some subcommands.)

---

> It's more subtle than that: Cobra does care, but only in one direction: it can accept a global option on the right-hand side, but cannot accept a local (right-hand) option on the left (before the subcommand).

It’s even _more_ subtle than that: it _sometimes_ works in the local → global direction as well:
```console
% bin/skopeo --src-creds=foo copy a b
# does not complain about --src-creds being unknown
```
and
```console
% bin/skopeo --unknown=a copy a b
FATA[0000] unknown flag: --unknown
```
vs.
```console
% bin/skopeo --unknown copy a b
FATA[0000] unknown command "a" for "skopeo"             
```

When interpreting the command line (without `TraverseChildren`), Cobra first looks for the (optionally nested) subcommand name by stripping flags from the command line; after it finds the subcommand, it merges all relevant flags into one `FlagSet` (somehow, perhaps with priorities), strips the recognized (nested) subcommand name from the command line, and then uses the merged `FlagSet` on the flags-only command line.

The subtlety is that dividing the command line into subcommands/flags requires the parser, in some cases, to know which flags require an argument, and the code happens to assume that unknown flags do require an argument; so `--unknown copy a b` is interpreted as `--unknown=copy a b`, which explains the error message. And _this_ part does not use merged flags, i.e. a local option is not interpreted correctly when put to the left of the subcommand name (but it may be accepted if the parser ends up identifying the subcommand).

Reply:
There’s even more subtleties to this, see https://github.com/spf13/cobra/issues/1449 .

But I think we might actually get the outcome we want, it seems: don’t use `PersistentFlags`, but a separate local flag for the root and for (all current) subcommands; and set `TraverseChildren`. If there isn’t an inherited persistent flag, all the ambiguity goes away.

Test program:
```go
package main

import (
	"bytes"
	"fmt"
	"strings"

	"github.com/spf13/cobra"
)

type options struct {
	deprecatedFlag bool
	sameNameFlag   bool
	srcFlag        bool
}

func dummyRun(_ *cobra.Command, _ []string) {}

func createApp() (*cobra.Command, *options) {
	opts := options{}

	rootCommand := &cobra.Command{Use: "main", Long: "Top-level main help"}
	rootCommand.Flags().BoolVar(&opts.deprecatedFlag, "flag", false, "deprecated - use per-subcommand flags")
	if err := rootCommand.Flags().MarkHidden("flag"); err != nil {
		panic(`unable to mark "flag" as hidden`)
	}
	rootCommand.PersistentFlags().BoolVar(&opts.sameNameFlag, "global", false, "foo")
	rootCommand.TraverseChildren = true

	sameNameCmd := &cobra.Command{Use: "same-name", Long: "same-name help", Run: dummyRun}
	sameNameCmd.Flags().BoolVar(&opts.sameNameFlag, "flag", false, "this is not deprecated")
	rootCommand.AddCommand(sameNameCmd)

	otherNameCmd := &cobra.Command{Use: "other-name", Long: "other-name help", Run: dummyRun}
	otherNameCmd.Flags().BoolVar(&opts.deprecatedFlag, "flag", false, "deprecated - use src-flag")
	if err := otherNameCmd.Flags().MarkHidden("flag"); err != nil {
		panic(`unable to mark "flag" as hidden`)
	}
	otherNameCmd.Flags().BoolVar(&opts.srcFlag, "src-flag", false, "this is not deprecated either")
	rootCommand.AddCommand(otherNameCmd)
	return rootCommand, &opts
}

func main() {
	for _, t := range []struct {
		args     []string
		expected options
	}{
		{[]string{"same-name"}, options{}}, // No flags
		{[]string{"same-name", "--flag"}, options{sameNameFlag: true}},

		{[]string{"--flag", "same-name"}, options{deprecatedFlag: true}},

		{[]string{"--flag", "other-name"}, options{deprecatedFlag: true}},
		{[]string{"other-name", "--flag"}, options{deprecatedFlag: true}},
		{[]string{"other-name", "--src-flag"}, options{srcFlag: true}},
	} {
		app, opts := createApp()
		app.SetArgs(t.args)
		if err := app.Execute(); err != nil {
			panic(fmt.Sprintf("Execute failed for %#v", t.args))
		}
		if *opts != t.expected {
			panic(fmt.Sprintf("Unexpected parsing results for %#v: %#v vs. %#v", t.args, opts, t.expected))
		}
	}

	for _, t := range []struct {
		args    []string
		visible bool
	}{
		{[]string{"--help"}, false},
		{[]string{"same-name", "--help"}, true}, // Would like to have true here
		{[]string{"other-name", "--help"}, false},
	} {
		app, _ := createApp()
		stdout := bytes.Buffer{}
		app.SetOut(&stdout)
		app.SetArgs(t.args)
		if err := app.Execute(); err != nil {
			panic(fmt.Sprintf("Execute failed for %#v", t.args))
		}
		help := stdout.String()
		fmt.Printf("===\n%s\n===\n", help)
		visible := strings.Contains(help, "--flag")
		if visible != t.visible {
			panic(fmt.Sprintf("Unexpected visibility for %#v: %v vs. %v", t.args, visible, t.visible))
		}
	}
}
```